jung-algorithms/src/main/java/edu/uci/ics/jung/layout/algorithms/TreeLayoutAlgorithm.java

```diff
@@ -152,54 +158,28 @@ private int calculateDimensionX(LayoutModel<N> layoutModel, Collection<N> roots)
     return size;
   }
 
-  protected void setCurrentPositionFor(LayoutModel<N> layoutModel, N node) {
-    int width = layoutModel.getWidth();
-    int height = layoutModel.getHeight();
-    int x = (int) this.currentX;
-    int y = (int) this.currentY;
-    if (x < 0) {
-      width -= x;
-    }
+  private int calculateDimensionY(LayoutModel<N> layoutModel, N node) {
 
-    if (x >= width - distX) {
-      width = x + distX;
-    }
+    int size = 0;
+    int childrenNum = layoutModel.getGraph().successors(node).size();
 
-    if (y < 0) {
-      height -= y;
-    }
-    if (y >= height - distY) {
-      height = y + distY;
-    }
-    if (layoutModel.getWidth() < width || layoutModel.getHeight() < height) {
-      layoutModel.setSize(width, height);
+    if (childrenNum != 0) {
+      for (N element : layoutModel.getGraph().successors(node)) {
+        size = Math.max(size, calculateDimensionY(layoutModel, element) + distX);
+      }
     }
-
-    //    Point location = layoutModel.get(node);
-    setLocation(layoutModel, node, this.currentX, this.currentY);
-  }
-
-  /**
-   * can be overridden to add behavior
-   *
-   * @param layoutModel
-   * @param node
-   * @param location
-   */
-  protected void setLocation(LayoutModel<N> layoutModel, N node, Point location) {
-    layoutModel.set(node, location);
+    return size;
   }
 
-  protected void setLocation(LayoutModel<N> layoutModel, N node, double x, double y) {
-    layoutModel.set(node, x, y);
-  }
+  private int calculateDimensionY(LayoutModel<N> layoutModel, Collection<N> roots) {
```

Comment:
nano-nit: `roots` could be `Iterable<N>`

Reply:
yes, but that's the way I found them. I am not the author of TreeLayoutAlgorithm